### PR TITLE
JBR-6696: hardened MTLGC_DestroyMTLGraphicsConfig 

### PR DESCRIPTION
**Pull request by bourgesl**
to ensure any aync callbacks are performed before deallocation, added CVDisplayLink checks, traces and refactored cleanup code + fixed review comments

### CODE REVIEW

**Comment by bourgesl** [comment]
Will fix PR asap

**Comment by avu** [comment]
Did you check for freezes that happened with the previous version?

**Comment by bourgesl** [comment]
Yes I am testing my patch ...

**Comment by bourgesl** [comment]
(WIP) tests in progress... 
still annoying trouble with (not direct surface) when scaling factor are changed and new frame created...

File: src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLContext.m [hunk 1]
```diff
@@ -39,6 +39,10 @@
 // scenarios with multiple subsequent updates.
 #define KEEP_ALIVE_COUNT 4
 
+// Min interval between 2 display link callbacks (Main thread may be blocked)
+// ~ 2ms (shorter than 240Hz frame time)
+#define KEEP_ALIVE_MIN_INTERVAL 2.0 / 1000.0
```
Comment (mkartashev) [inline src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLContext.m:44]: Maybe rephrase it in terms of FPS because it seems to better describe this restriction. Something like
```C
#define MAX_FPS 240
#define KEEP_ALIVE_MIN_INTERVAL 1.0/MAX_FPS
```

Comment (bourgesl) [reply src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLContext.m:44]: Yes, it seems possible to query monitor capabilities at runtime on recent macos, so yes, max_freq or fps is important as new monitors can achieve 480hz...
Here I just want to discard old callbacks  ... postponed by busy main thread, then flushed at once...
Will look at cvtime ...

File: src/java.base/macosx/native/libjli/java_md_macosx.m [hunk 2]
```diff
@@ -288,6 +288,52 @@
...
+
+    // disabling may report wrong positives (in apple code):
+    setenv("MTC_SUPPRESS_SYSTEM_REPORTS", "1", 1);
+
+    if (0) {
+        // selector stats:
+        setenv("MTC_PRINT_SELECTOR_STATS", "1", 1);
+    }
+
+    void *result = dlopen("/Applications/Xcode.app/Contents/Developer/usr/lib/libMainThreadChecker.dylib",
```
Comment (avu) [inline src/java.base/macosx/native/libjli/java_md_macosx.m:327]: I think it's better to provide the path via JVM option

File: src/java.desktop/macosx/native/libosxapp/ThreadUtilities.m [hunk 3]
```diff
@@ -135,6 +167,67 @@ + (void)performOnMainThread:(SEL)aSelector on:(id)target withObject:(id)arg wait
...
+        return;
+    }
+
+    // Perform instrumentation on selector:
+    static mach_timebase_info_data_t* timebase = nil;
+    if (timebase == nil) {
+        timebase = malloc(sizeof (mach_timebase_info_data_t));
+        mach_timebase_info(timebase);
+
+        printf("timebase->numer: %d\n", timebase->numer);
```
Comment (avu) [inline src/java.desktop/macosx/native/libosxapp/ThreadUtilities.m:179]: Maybe it's better to use J2D runtime trace here (J2dRls...)

File: src/java.desktop/macosx/native/libosxapp/ThreadUtilities.m [hunk 4]
```diff
@@ -110,20 +120,42 @@ + (void)invokeBlockCopy:(void (^)(void))blockCopy {
...
+    return nil;
+}
+
 + (void)performOnMainThread:(SEL)aSelector on:(id)target withObject:(id)arg waitUntilDone:(BOOL)wait {
-    if ([NSThread isMainThread] && wait == YES) {
-        [target performSelector:aSelector withObject:arg];
-    } else {
-        if (wait && isEventDispatchThread()) {
-            void (^blockCopy)(void) = Block_copy(^(){
+    if (!DECORATE_MAIN_THREAD) {
```
Comment (avu) [inline src/java.desktop/macosx/native/libosxapp/ThreadUtilities.m:154]: We use DECORATE_MAIN_THREAD as a macro so it's better to replace the if statement with the preprocessor IF or maybe enable it always in DEBUG mode.

**Comment by bourgesl** [comment]
New PR opened: https://github.com/JetBrains/JetBrainsRuntime/pull/383
